### ISSUE TITLE
Requirements and specs for end-2-end encryption should be clearly defined

### ISSUE THREAD

**Issue by evilaliv3** [globaleaks/GlobaLeaks#1153]
With this ticket i wanna open the discussion on the requirements we want globaleaks to satisfy when implementing the end-2-end encryption. this in order to organize the work and performing a complete  security analysis and planning the development ticket by ticket.

To this aim this initial message will act as a resume of the discussion with reference to all the different tickets we will open and address working on this topic.

Before starting opening tons of tickets let's us this one as bitchy ticket in order to discuss it all and then i will open the child ones.

Requirements not so tricky that i consider quite all useful and staightforward and needed as first step in order to improve the system without adding new capabilities; could be considered the first step into the integration: The only one really critical for usability is 8 and we should take care to discuss that in particular
* [ ] 1 The interface should enable the upload of  ascii armored pgp keys;
* [ ] 2 The interface should enable the upload an also binary pgp keys (load only, needed in order to support preexisting keys but globaleaks will always export ascii armored keys)
* [ ] 3 The interface should always provide/get ascii armored pgp keys to/from the server
* [ ] 4 The interface should be able to validate uploaded keys
* [ ] 4spec1 The uploaded keys should be validated before been uploaded
* [ ] 4spec2 The client should verify that the key is encrypted with the key of the user (cached on login) and send it to the server only if this constraint is met.
* [ ] 5 The interface should enable to generate a key from the interface by simply clicking a button
* [ ] 5spec1 Generated keys through the interface should be mapped on the receiver mail (ino order to minimize user input)
* [ ] 5spec2 Generated keys through the interface should have a 4096 private key (with no choice for the user; we should use automagical secure defaults at least in first implementations) (no order to minimize user input)
* [ ] 5spec3 The private key should be protected with the password of the receiver; the password will be cached by the glclient on the login of the user in order to be automatically availably onto PGP key generation and on crypto operations by means of the PGP key.
* [ ] 6 Generated keys should be simply downloadable by the user in the most usable common and usable format. we should evalutate firtst the possibility to boundle them as a keyring (openpgp.js offer it) and see if this keyring is usable on every OSs by performing a single click. other solution eventually involve the possibility to download the pgp keys as zip file containing both public.asc and private.asc

Requirements needed in order to go forward the current work:
* [ ] 7 The interface should encrypt fields of the submission for each receiver of the submission (single encryption); this involve discussing if we should have a single encryption or independent encryptions for each receiver. in addition we should think if we need to have a single encryption of the submission or a per field encryption and the benefit of the two approach
* [ ] 8 The interface should encrypt files of the submission for each receiver of the submission (single encryption); this involve discussing if we should have a single encryption or independent encryptions for each receiver. quite for sure here we should go for the single encryption in order to minimize the network transfers (1GBx10 = 10GB! maths! arg)
* [ ] 9 The interface should generate automatically a key for the whistleblower and give the wistleblower a secret in order to recreate the key deterministically when he returns onto the plaform
* [ ] 10 Upon submission the public key of the whistleblower should be saved together with the submission
* [ ] 11 I know that 9 wont be easy replicable also for receiver due to the different validity of the keys but probably we should think to a similar solution also for receivers: how to avoid that the receiver should load his key and password of the key on the platform any time he need to access the submissions?

this is simply the start; i've a plethora of ideas but this should be enough to start asking your comments:
@hellais @mmaker @vecna @fpietrosanti @origliante

**Comment by fpietrosanti** [comment]
Point 6-7-8 are specs of the functional requirement 5

**Comment by evilaliv3** [comment]
i agree but now we do not need to go for the subtle but try to focus on all the ideas and then clearly organize it in a hierarchical way. various of the requirements i've written ar most spec, but given that that globaleaks want to simplify the user adoption that spec appear as a requirement.

i've renamed 6-7-8 as 5spec1, 5spec2, 5spec3 and explicited the intent and why i consider them quite a requirement and not simply a spec.

**Comment by vecna** [comment]
Point 2: I never see binary keys around. I feel keyserver integration can be better, keybase.io integration, a little premature, but interesting too.

Point 3: you've to think that a normal user never open a terminal to do gpg --import, so we've to figure out which kind of capability we can guarantee, and specify accordingly. I guess, the point three, means "Interface should be able to extract fingerprint, basic info, from the retrieved PGP keys" because can be useful, for WB safety/paranoia, dump that information if want check keys offiline.

Point 4: for "validate" you mean, import and check format integrity ? 

Point 5spec3: what about, using the password of the globaleaks access, just hashed in a different way ? 

Point 6 and 1 are face of the same coin, right ? the user dump the keys, later can re-upload and keep them stored in the server, later can download them again, etc. am I right ? 

Point 10: we can copy the code from miniLock, do exactly that https://github.com/kaepora/miniLock (from a secret, generate a key pair), but I guess we can just use the receipt, hashed in a different way again. keeping the receipt as "the only secret" in the hand of the WB.

Point 11: agree, because is part of the receipt saving.

**Comment by evilaliv3** [comment]
Concerning point 2 you are right probably we can skip that requirement, or make it as secondari ticket in backlog.
Concerning Point 3: you are right. i think that we should rely on the operating system binding and look if there is a particular extensions that is mostly recognized as a PGP key to be imported on the various operating systems. on TAILS i think a double click on an .ASC file will make the trick, but what about other operating systems?

Point 3: exactly i mean that when a key is loaded (for example when now it is pasted) the client should rely on openpgp.js to do a prevalidation, prepare the armor if the key was binary key, print out the information about the loaded key and the expiration date and permit to uploaded it effectively only if the key is loaded. this is needed in order to provide a user feedback to the user on the action he is taking also before that the user effectively click to save his preferences.

Point 5spec: you are great man, this seems an interesting idea to discuss it's ok?the only problem i see would be when the user will have the need to change is own password. he will then forgot his own password so probably i think the best initial choice would be to ask for a new password and delegate to an improval ticket to solve the issue.

Point 6 and 1: in general in my thinking the system would always need to store only the public key, and the receiver/admin would have the possibility to generate/regenerate a key. on such event the private/public key bundle should be downloadable clientside, and the public key only would be pushed upstream to globaleaks in an ascii armored way exactly as we are doing now.

Point 10: this is exactly what i'm thinking but for the moment let's concentrate more on requirements in order to define the research directions.

Point 11: yep the idea of the zip is to simplify at least in the initial step for the setup. the user get a single file that contains both the private and the public key already separated in to files private.asc and public.asc

@vecna apart form what i've written do you foresee any other requirement/spec/idea; i wanna this ticket to be abused by a fucking brainstorm!

i've quite all the workflow in mind except for the only great problem i see, that is: how to avoid that every time the user access the platform he has to clientside load his application key?

**Comment by fpietrosanti** [comment]
D2 original planning: https://docs.google.com/document/d/1pDe_s7875KJPg3IVYpWoxnyAP8GqOnWn-3OuaEjulvo/edit?usp=sharing

**Comment by fpietrosanti** [comment]
Working document for detailed specs https://docs.google.com/document/d/1ShdxubexlFPKedhO28i0RvnjHSiQU4lma5B0DSs2xsQ/edit?usp=sharing

**Comment by fpietrosanti** [comment]
See Arturo’s technical analysis proposal of September 2014 http://lists.globaleaks.org/pipermail/globaleaks-dev/2014-September/000008.html

**Comment by hellais** [comment]
What is the ordering that you have given to these issues?

Here is some information on what has already been done WRT implementing some of what is described above.

9:
Implementing this requires changes both in the file upload client, but also in the file upload handler backend.
Changes needed to make this work with cyclone can be found here: https://github.com/fiorix/cyclone/pull/144.

I also wrote some hooks that would allow preprocessing of the chunks to be uploaded by jquery file uploader: https://github.com/hellais/jQuery-File-Upload/tree/feature/upload-handler. I believe this solution is very hackish, mainly due to the fact that the jquery file upload code is quite messy.

I have been looking into some other solutions and I have found flow.js (https://github.com/flowjs/flow.js). It's code is very clean and it natively supports a chunk `preprocessor` that would work well for our use-case. They also have a very well written angular.js extension (https://github.com/flowjs/ng-flow).

Is anybody strongly against switching to flow.js from jquery file uploader?

The other piece of this puzzle is having openpgp.js support streaming. I implemented this here: https://github.com/openpgpjs/openpgpjs/pull/260, but I need to make some changes to the libraries I use in order to have it merged into openpgp.js (in particular I must replace the SHA256 implementation with that of Forge and not use node buffers) 

10:
To generate deterministically the whistleblower key here you may find the code available here: https://github.com/openpgpjs/openpgpjs/pull/256, but it didn't get merged.

The reason why this can't be done, as @vecna suggested, with MiniLock is that MiniLock does not use the OpenPGP protocol.
What we can do, however, is use a similar approach to theirs in the sense that we take a short pre shared secret and stretch it with scrypt and apply that as a seed to the PRNG in the above code.

**Comment by evilaliv3** [comment]
hey art! right know we have not assigned any order to tickets we are still brainstorming and a lot of the things i've written are wrong cause i was forgotting old discussions and decisions that we already written in the specs reported by @fpietrosanti. i'm  now going to correct the ticket message and i will write a new comment when done.

**Comment by fpietrosanti** [comment]
@hellais I'm against switching from jquery file uploader because:
* We already have several bit of code (and knowledge) using it
* Is widely used and supported (4912 fork of JQFU vs. 118 fork of flow.js)
* The indirect community impact of our integration with OpenPGP.js/JQFU is much higher

**Comment by fpietrosanti** [comment]
In general i think that we need to finalize the overall technical specifications, including the consideration for ciphers, algorithms, entropy bits, encrypted data formats

**Comment by hellais** [comment]
@fpietrosanti the effort of getting the patches we need into jquery file uploader is though much greater than switching to another library.

Also have you looked at the code I posted above? It's terrible and we should not be relying on that, though I don't see any better way of doing it with the existing jquery file uploader code base.

**Comment by evilaliv3** [comment]
@fpietrosanti i'm not totally against the switch to flow.js for the following reasons:

due to jquery file uploader we are now forced to include various libraries jquery, jquery-ui with a lot of unneded js and css. if flow.js is small and self containded and works it would be good to switch to it.

from feature point of view it does not lacks nothing and it seems that it's going to be more maintained and integrated with angular.
a feature missing probably is the file preview that anyhow is causing issues on TBB so probably we sholuld remove it for now (https://github.com/globaleaks/GlobaLeaks/issues/1147).

**Comment by hellais** [comment]
@evilaliv3 actually the angular.js extension for flow supports image preview without using the HTML5 canvas. See: https://github.com/flowjs/ng-flow#how-can-i-preview-uploaded-image.

**Comment by vecna** [comment]
About flow.js vs JQFU: we've to consider that end2end encryption is so big that probably we've to make a dedicated set of intermediary releases with it. also if flow.js has something less than JQFU, we'll figure out gradually.

@evilaliv3 point5 spec: client side has to be managed the "password change" with the proper communication, and request of "put your SD with private key" otherwise you don't change password ? and we enforce client side that the password is always 1.. or we can let user keep two password if (s)he want ? 

point zip: means support zip in javascript, I was suggesting a planintext file, with some comment, public key, private key (encrypted), both armored, and if we like integrity, the file is signed with the node privkey.

**Comment by evilaliv3** [comment]
i agree totally with vecna.
@hellais for the moment lets proceed with start doing it with flow.js that is most important for us. are you restarting working on this topic?

@vecna point 5 of the spec i did a mistake, i'm now reviewing it all. i was forgetting that we choosed to adopt the hushmail model so that the key will be stored encrypted on the server with no need for the user to load it everytime. the key will always be encrypted with the same password of the platform and on password change only the key should be automagically re-encrypted and loaded on the server.

**Comment by evilaliv3** [comment]
@vecna ok i've updated 5spec3 and 6

**Comment by evilaliv3** [comment]
@hellais now that you are starting reworking on this can you align with @origliante to which i've assigned some of the starting activities on key generation and imports?

as @fpietrosanti suggested we won't add so much improvement in doing an early integration of the key generation feature i was suggesting (apart from making possible to speed up some setups and trainings. so for the moment let's only concentrate on the medium term goal that is to have e2e encryption all in place.
to this aim for example requirements above generally will involve loading/exporting the private key (se requirement 1) cause when one will need to load a key on the system he will need to load the ancrypted private key (that is needed be kept on the server.
to this aim i've added 4spec1 and 4spec2; please see them.

**Comment by evilaliv3** [comment]
this ticket should be reformulated in relation to the specs writtend on https://docs.google.com/document/d/1ShdxubexlFPKedhO28i0RvnjHSiQU4lma5B0DSs2xsQ/edit?usp=sharing . its best to switch using that document to keep track of the various changes and then cleanup this ticket with the exact things to be implemented.

**Comment by evilaliv3** [comment]
closing this as superseaded by the ideas developed during the hackaton so that this ticket represent simply a mess.
